File: backend/src/test/java/com/inventory/service/RawMaterialServiceTest.java
package com.inventory.service;

import com.inventory.dto.RawMaterialDTO;
import com.inventory.entity.RawMaterial;
import com.inventory.exception.ResourceNotFoundException;
import com.inventory.repository.RawMaterialRepository;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import java.util.Arrays;
import java.util.List;
import java.util.Optional;

import static org.junit.jupiter.api.Assertions.*;
import static org.mockito.ArgumentMatchers.any;
import static org.mockito.Mockito.*;

@ExtendWith(MockitoExtension.class)
class RawMaterialServiceTest {
    
    @Mock
    private RawMaterialRepository rawMaterialRepository;
    
    @InjectMocks
    private RawMaterialService rawMaterialService;
    
    private RawMaterial rawMaterial;
    
    @BeforeEach
    void setUp() {
        rawMaterial = new RawMaterial();
        rawMaterial.setId(1L);
        rawMaterial.setName("Steel");
        rawMaterial.setStockQuantity(1000);
    }
    
    @Test
    void testFindAll() {
        List<RawMaterial> rawMaterials = Arrays.asList(rawMaterial);
        when(rawMaterialRepository.findAll()).thenReturn(rawMaterials);
        
        List<RawMaterialDTO> result = rawMaterialService.findAll();
        
        assertNotNull(result);
        assertEquals(1, result.size());
        assertEquals("Steel", result.get(0).getName());
        verify(rawMaterialRepository, times(1)).findAll();
    }
    
    @Test
    void testFindById_Success() {
        when(rawMaterialRepository.findById(1L)).thenReturn(Optional.of(rawMaterial));
        
        RawMaterialDTO result = rawMaterialService.findById(1L);
        
        assertNotNull(result);
        assertEquals("Steel", result.getName());
        assertEquals(1000, result.getStockQuantity());
    }
    
    @Test
    void testFindById_NotFound() {
        when(rawMaterialRepository.findById(999L)).thenReturn(Optional.empty());
        
        assertThrows(ResourceNotFoundException.class, () -> {
            rawMaterialService.findById(999L);
        });
    }
    
    @Test
    void testCreate() {
        RawMaterialDTO rawMaterialDTO = new RawMaterialDTO();
        rawMaterialDTO.setName("Aluminum");
        rawMaterialDTO.setStockQuantity(500);
        
        when(rawMaterialRepository.save(any(RawMaterial.class))).thenReturn(rawMaterial);
        
        RawMaterialDTO result = rawMaterialService.create(rawMaterialDTO);
        
        assertNotNull(result);
        verify(rawMaterialRepository, times(1)).save(any(RawMaterial.class));
    }
    
    @Test
    void testUpdate_Success() {
        RawMaterialDTO rawMaterialDTO = new RawMaterialDTO();
        rawMaterialDTO.setName("Updated Steel");
        rawMaterialDTO.setStockQuantity(1500);
        
        when(rawMaterialRepository.findById(1L)).thenReturn(Optional.of(rawMaterial));
        when(rawMaterialRepository.save(any(RawMaterial.class))).thenReturn(rawMaterial);
        
        RawMaterialDTO result = rawMaterialService.update(1L, rawMaterialDTO);
        
        assertNotNull(result);
        verify(rawMaterialRepository, times(1)).save(any(RawMaterial.class));
    }
    
    @Test
    void testDelete_Success() {
        when(rawMaterialRepository.existsById(1L)).thenReturn(true);
        
        rawMaterialService.delete(1L);
        
        verify(rawMaterialRepository, times(1)).deleteById(1L);
    }
    
    @Test
    void testDelete_NotFound() {
        when(rawMaterialRepository.existsById(999L)).thenReturn(false);
        
        assertThrows(ResourceNotFoundException.class, () -> {
            rawMaterialService.delete(999L);
        });
    }
}

File: backend/src/test/kotlin/com/inventory/service/RawMaterialServiceTest.kt
package com.inventory.service

import com.inventory.dto.RawMaterialDTO
import com.inventory.entity.RawMaterial
import com.inventory.repository.RawMaterialRepository
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.mockito.InjectMocks
import org.mockito.Mock
import org.mockito.Mockito.*
import org.mockito.junit.jupiter.MockitoExtension
import java.util.*

@ExtendWith(MockitoExtension::class)
class RawMaterialServiceTest {

    @Mock
    private lateinit var rawMaterialRepository: RawMaterialRepository

    @InjectMocks
    private lateinit var rawMaterialService: RawMaterialService

    private lateinit var testRawMaterial: RawMaterial
    private lateinit var testDTO: RawMaterialDTO

    @BeforeEach
    fun setup() {
        testRawMaterial = RawMaterial(
            id = 1L,
            name = "Farinha",
            stockQuantity = 100
        )

        testDTO = RawMaterialDTO(
            id = 1L,
            name = "Farinha",
            stockQuantity = 100
        )
    }

    @Test
    fun `should find all raw materials`() {
        // Given
        val rawMaterials = listOf(testRawMaterial)
        `when`(rawMaterialRepository.findAll()).thenReturn(rawMaterials)

        // When
        val result = rawMaterialService.findAll()

        // Then
        assertNotNull(result)
        assertEquals(1, result.size)
        assertEquals("Farinha", result[0].name)
        verify(rawMaterialRepository).findAll()
    }

    @Test
    fun `should find raw material by id`() {
        // Given
        `when`(rawMaterialRepository.findById(1L)).thenReturn(Optional.of(testRawMaterial))

        // When
        val result = rawMaterialService.findById(1L)

        // Then
        assertNotNull(result)
        assertEquals(1L, result.id)
        assertEquals("Farinha", result.name)
        assertEquals(100, result.stockQuantity)
        verify(rawMaterialRepository).findById(1L)
    }

    @Test
    fun `should create new raw material`() {
        // Given
        val newRawMaterial = RawMaterial(
            name = "Açúcar",
            stockQuantity = 50
        )
        val savedRawMaterial = newRawMaterial.copy(id = 2L)
        
        `when`(rawMaterialRepository.save(any(RawMaterial::class.java)))
            .thenReturn(savedRawMaterial)

        // When
        val dto = RawMaterialDTO(name = "Açúcar", stockQuantity = 50)
        val result = rawMaterialService.create(dto)

        // Then
        assertNotNull(result)
        assertEquals(2L, result.id)
        assertEquals("Açúcar", result.name)
        assertEquals(50, result.stockQuantity)
        verify(rawMaterialRepository).save(any(RawMaterial::class.java))
    }

    @Test
    fun `should update existing raw material`() {
        // Given
        `when`(rawMaterialRepository.findById(1L)).thenReturn(Optional.of(testRawMaterial))
        `when`(rawMaterialRepository.save(any(RawMaterial::class.java)))
            .thenReturn(testRawMaterial.apply { stockQuantity = 150 })

        // When
        val updateDTO = RawMaterialDTO(id = 1L, name = "Farinha", stockQuantity = 150)
        val result = rawMaterialService.update(1L, updateDTO)

        // Then
        assertNotNull(result)
        assertEquals(1L, result.id)
        assertEquals(150, result.stockQuantity)
        verify(rawMaterialRepository).findById(1L)
        verify(rawMaterialRepository).save(any(RawMaterial::class.java))
    }

    @Test
    fun `should delete raw material`() {
        // Given
        `when`(rawMaterialRepository.existsById(1L)).thenReturn(true)

        // When
        rawMaterialService.delete(1L)

        // Then
        verify(rawMaterialRepository).existsById(1L)
        verify(rawMaterialRepository).deleteById(1L)
    }
}
